pkg/admission/permissionclaims/mutating_permission_claims.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaims
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/permissionclaims"
+)
+
+const (
+	byWorkspaceIndex = "mutating-permission-byWorkspace"
+	PluginName       = "apis.kcp.dev/PermissionClaims"
+)
+
+type mutatingPermissionClaims struct {
+	*admission.Handler
+
+	apiBindingsIndexer cache.Indexer
+	apiExportIndexer   cache.Indexer
+	bindingReady       func() bool
+	exportReady        func() bool
+}
+
+var _ admission.MutationInterface = &mutatingPermissionClaims{}
+
+func NewMutatingPermissionClaims() admission.MutationInterface {
+
+	p := &mutatingPermissionClaims{}
+	p.Handler = admission.NewHandler(admission.Create)
+	p.SetReadyFunc(func() bool {
+		if p.bindingReady() && p.exportReady() {
+			return true
+		}
+		return false
+	})
+	return p
+}
+
+func (m *mutatingPermissionClaims) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	bindings, err := m.apiBindingsIndexer.ByIndex(byWorkspaceIndex, lcluster.String())
+	if err != nil {
+		return err
+	}
+
+	for _, b := range bindings {
```

Comment:
plus core types

Reply:
Binding in a given cluster would have an ObservedAcceptedPermissionClaims with a core type; if that exists, it would be labeled. So we don't have to loop through core types here because nothing would be labeled, right, unless there is a binding, right?

Reply:
true